File: 2020/day05-1.hs
type Codes = (Int, Int)

solution :: String -> Int
solution = maximum . map (codesToSeatId . parseLine) . lines

codesToSeatId :: Codes -> Int
codesToSeatId (a, b) = a * 8 + b

parseLine :: String -> Codes
parseLine line = (
    toInt . map (\x -> (if' (x == 'B') 1 0)) . take 7 $ line,
    toInt . map (\x -> (if' (x == 'R') 1 0)) . drop 7 $ line)

toInt :: [Int] -> Int
toInt [] = 0
toInt list = helper 1 (reverse list)
    where
        helper _ [] = 0
        helper n (1:xs) = n + helper (2*n) xs
        helper n (_:xs) = helper (2*n) xs

if' :: Bool -> a -> a -> a
if' True a _ = a
if' False _ a = a

main :: IO ()
main = do readFile "day05.dat" >>= print . solution

File: 2020/day05-1.ts
function solution(content: string) {
  return Math.max(...content.split("\n").map((line) => parse(line)));
}

function parse(line: string) {
  const row = toInt(
    line
      .slice(0, 7)
      .split("")
      .map((x) => (x === "B" ? 1 : 0))
  );
  const col = toInt(
    line
      .slice(7)
      .split("")
      .map((x) => (x === "R" ? 1 : 0))
  );
  return row * 8 + col;
}

function toInt(binary: number[]) {
  function helper(mult: number, list: number[]): number {
    if (list.length === 0) return 0;
    const [first, ...rest] = list;
    if (first === 1) return mult + helper(2 * mult, rest);
    return helper(2 * mult, rest);
  }
  return helper(1, binary.slice().reverse());
}

Deno.readTextFile("day05.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
